### PR TITLE
Show `stdout` and `stderr` for linter bears

### PR DESCRIPTION
**Pull request by yash-nisar**
Print the `stdout` and `stderr` from linter bears when tests
fail which is helpful for debugging.

Closes https://github.com/coala/coala-bears/issues/455

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by yash-nisar** [comment]
CC @Makman2

File: coalib/testing/LocalBearTestHelper.py [hunk 1]
```diff
@@ -12,16 +13,38 @@
 
 
 @contextmanager
-def execute_bear(bear, *args, **kwargs):
+def execute_bear(bear, *args, **kwargs):  # pragma: no cover
```
Comment (Nosferatul) [inline coalib/testing/LocalBearTestHelper.py]: why no cover?

Comment (yash-nisar) [inline coalib/testing/LocalBearTestHelper.py]: This patch involves printing the `stderr` and `stdout` from linter bears when there is some problem or an exception is raised. To test this feature, I will deliberately have to raise an exception and thus make the tests fail so as to see if we're printing the `stderr` and `stdout` or not and IIRC once, we raise an exception, the tests hault there itself and we wouldn't be able to proceed further. However, I've [tested this manually](https://github.com/coala/coala-bears/issues/455#issuecomment-317270508). :smile:

Comment (Nosferatul) [inline coalib/testing/LocalBearTestHelper.py]: nice!

File: coalib/testing/LocalBearTestHelper.py [hunk 2]
```diff
@@ -12,16 +13,38 @@
...
+            bear_output_generator = bear.execute(*args, **kwargs)
+
         assert bear_output_generator is not None, \
             'Bear returned None on execution\n'
         yield bear_output_generator
     except Exception as err:
         msg = []
         while not bear.message_queue.empty():
             msg.append(bear.message_queue.get().message)
+        msg.extend(console_output)
```
Comment (Nosferatul) [inline coalib/testing/LocalBearTestHelper.py]: this is needed?

Comment (Makman2) [inline coalib/testing/LocalBearTestHelper.py]: yes we want to append the messages to the assertion error. This is the same as doing `msg += console_output`.

File: coalib/testing/LocalBearTestHelper.py [hunk 3]
```diff
@@ -14,14 +15,36 @@
 @contextmanager
 def execute_bear(bear, *args, **kwargs):
     try:
-        bear_output_generator = bear.execute(*args, **kwargs)
+        console_output = []
+
+        with ExitStack() as stack:
+            if hasattr(bear, 'process_output'):
```
Comment (Makman2) [inline coalib/testing/LocalBearTestHelper.py]: maybe a comment stating
```
# For linters provide additional information, such as stdout and stderr.
```

File: coalib/testing/LocalBearTestHelper.py (hunk 2 above)
Comment (Makman2) [inline coalib/testing/LocalBearTestHelper.py]: @yash-nisar maybe you should really do `msg += console_output`, this is kinda easier to read :+1:

File: tests/testing/LocalBearTestHelperTest.py [hunk 4]
```diff
@@ -101,8 +104,44 @@ def setUp(self):
...
+            with execute_bear(uut, 'filename', ['file']) as result:
+                self.assertNotEqual(result, [])
+        except AssertionError as ex:
+            self.assertIn('hello stdout', str(ex))
+            self.assertNotIn('hello stderr', str(ex))
 
+    def test_exception(self):
         with self.assertRaises(AssertionError), execute_bear(
-                self.uut,  'Luke', files[0]) as result:
+                self.uut, 'Luke', files[0]) as result:
```
Comment (Makman2) [inline tests/testing/LocalBearTestHelperTest.py]: 1. unnecessary change
2. did something change in the line with `def test_exception(self):`? Wondering why the diff removes it above and adds it here again :O

File: tests/testing/LocalBearTestHelperTest.py [hunk 5]
```diff
@@ -101,8 +104,44 @@ def setUp(self):
...
+        # Testing with both stdout and stderr enabled
+        uut = (linter(sys.executable, use_stdout=True, use_stderr=True)
+               (TestLinter)
+               (Section('TEST_SECTION'), Queue()))
+        try:
+            with execute_bear(uut, 'filename', ['file']) as result:
+                self.assertNotEqual(result, [])
+        except AssertionError as ex:
+            self.assertIn('hello stdout', str(ex))
+            self.assertIn('hello stderr', str(ex))
```
Comment (Makman2) [inline tests/testing/LocalBearTestHelperTest.py]: Could you also check if the string `Stdout:` is available? (same for `Stderr`)

File: tests/testing/LocalBearTestHelperTest.py (hunk 4 above)
Comment (yash-nisar) [inline tests/testing/LocalBearTestHelperTest.py]: Fixed :+1:

File: tests/testing/LocalBearTestHelperTest.py (hunk 5 above)
Comment (yash-nisar) [inline tests/testing/LocalBearTestHelperTest.py]: Sure :+1:

File: tests/testing/LocalBearTestHelperTest.py [hunk 6]
```diff
@@ -101,8 +104,45 @@ def setUp(self):
         section.append(Setting('exception', True))
         self.uut = TestBear(section, Queue())
 
-    def test_exception(self):
+    def test_stdout_stderr_on_linter_test_fail(self):
+        process_output_mock = Mock()
```
Comment (Makman2) [inline tests/testing/LocalBearTestHelperTest.py]: you don't use the mock, you can remove it

File: tests/testing/LocalBearTestHelperTest.py [hunk 7]
```diff
@@ -101,8 +104,45 @@ def setUp(self):
...
+
+        class TestLinter:
+            @staticmethod
+            def process_output(output, filename, file):
+                process_output_mock(output, filename, file)
```
Comment (Makman2) [inline tests/testing/LocalBearTestHelperTest.py]: and here you can just do nothing probably (so use `pass`)

File: tests/testing/LocalBearTestHelperTest.py (hunk 6 above)
Comment (yash-nisar) [inline tests/testing/LocalBearTestHelperTest.py]: Cool :+1:

Comment (yash-nisar) [inline tests/testing/LocalBearTestHelperTest.py]: My bad :(

File: tests/testing/LocalBearTestHelperTest.py (hunk 7 above)
Comment (yash-nisar) [inline tests/testing/LocalBearTestHelperTest.py]: Done :+1:

File: tests/testing/LocalBearTestHelperTest.py [hunk 8]
```diff
@@ -101,6 +103,45 @@ def setUp(self):
...
+            self.assertIn('Stderr:', str(ex))
+            self.assertIn('hello stderr', str(ex))
+
+        # Testing with only stdout enabled
+        uut = (linter(sys.executable, use_stdout=True)
+               (TestLinter)
+               (Section('TEST_SECTION'), Queue()))
+        try:
+            with execute_bear(uut, 'filename', ['file']) as result:
+                self.assertNotEqual(result, [])
```
Comment (Makman2) [inline tests/testing/LocalBearTestHelperTest.py]: actually, just raise an `AssertionError` :D That is easier to understand :) (same above)

**Comment by Makman2** [comment]
ack c1c8206

**Comment by Makman2** [comment]
Due to timeline and build problems: https://github.com/coala/coala/pull/4554 takes over.
